lib/web_ui/lib/src/engine/surface/recording_canvas.dart

```diff
@@ -283,41 +391,49 @@ class RecordingCanvas {
     _hasArbitraryPaint = true;
     _didDraw = true;
     final double paintSpread = _getPaintSpread(paint);
+    final PaintDrawDRRect command = PaintDrawDRRect(outer, inner, paint.paintData);
     _paintBounds.growLTRB(
       outer.left - paintSpread,
       outer.top - paintSpread,
       outer.right + paintSpread,
       outer.bottom + paintSpread,
+      command,
     );
-    _commands.add(PaintDrawDRRect(outer, inner, paint.paintData));
+    _commands.add(command);
   }
 
   void drawOval(ui.Rect rect, SurfacePaint paint) {
+    assert(!_debugRecordingEnded);
     _hasArbitraryPaint = true;
     _didDraw = true;
     final double paintSpread = _getPaintSpread(paint);
+    final PaintDrawOval command = PaintDrawOval(rect, paint.paintData);
     if (paintSpread != 0.0) {
-      _paintBounds.grow(rect.inflate(paintSpread));
+      _paintBounds.grow(rect.inflate(paintSpread), command);
     } else {
-      _paintBounds.grow(rect);
+      _paintBounds.grow(rect, command);
     }
-    _commands.add(PaintDrawOval(rect, paint.paintData));
+    _commands.add(command);
   }
 
   void drawCircle(ui.Offset c, double radius, SurfacePaint paint) {
+    assert(!_debugRecordingEnded);
     _hasArbitraryPaint = true;
     _didDraw = true;
     final double paintSpread = _getPaintSpread(paint);
+    final PaintDrawCircle command = PaintDrawCircle(c, radius, paint.paintData);
     _paintBounds.growLTRB(
       c.dx - radius - paintSpread,
```

Comment:
suggestion: final double distance = radius + paintSpread; and then use it 4 times...

Reply:
Done.